flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamRetractionRule.scala

```diff
@@ -0,0 +1,341 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.plan.nodes.datastream
+
+import org.apache.calcite.plan.{RelOptRule, RelOptRuleCall, RelTraitSet}
+import org.apache.calcite.plan.RelOptRule._
+import org.apache.calcite.plan.hep.HepRelVertex
+import org.apache.calcite.rel.RelNode
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ListBuffer
+
+/**
+  * Collection of retraction rules that apply various transformations on DataStreamRel trees.
+  * Currently, there are three transformations: InitProcessRule, NeedToRetractProcessRule and
+  * AccModeProcessRule. Note: these rules must be called in order (InitProcessRule ->
+  * NeedToRetractProcessRule -> AccModeProcessRule).
+  */
+object DataStreamRetractionRule {
+
+  /**
+    * Singleton rule that init retraction trait inside a [[DataStreamRel]]
+    */
+  val INIT_INSTANCE = new InitProcessRule()
+
+  /**
+    * Singleton rule that decide needToRetract property inside a [[DataStreamRel]]
+    */
+  val NEEDTORETRACT_INSTANCE = new NeedToRetractProcessRule()
+
+  /**
+    * Singleton rule that decide accMode inside a [[DataStreamRel]]
+    */
+  val ACCMODE_INSTANCE = new AccModeProcessRule()
+
+  /**
+    * Get all child RelNodes of a RelNode
+    * @param topRel The input RelNode
+    * @return All child nodes
+    */
+  def getChildRelNodes(topRel: RelNode): ListBuffer[RelNode] = {
+    val topRelInputs = new ListBuffer[RelNode]()
+    topRelInputs.++=(topRel.getInputs.asScala)
+    topRelInputs.transform(e => e.asInstanceOf[HepRelVertex].getCurrentRel)
+  }
+
+  def traitSetContainNeedToRetract(traitSet: RelTraitSet): Boolean = {
+    val retractionTrait = traitSet.getTrait(RetractionTraitDef.INSTANCE)
+    if (null == retractionTrait) {
+      false
+    } else {
+      retractionTrait.getNeedToRetract
+    }
+  }
+
+
+  /**
+    * Find all needToRetract nodes. A node needs to retract means that there are downstream
+    * nodes need retraction from it. Currently, [[DataStreamOverAggregate]] and
+    * [[DataStreamGroupWindowAggregate]] need retraction from upstream nodes, besides, a
+    * needToRetract node also need retraction from it's upstream nodes.
+    */
+  class NeedToRetractProcessRule extends RelOptRule(
+    operand(
+      classOf[DataStreamRel], none()),
+    "NeedToRetractProcessRule") {
+
+    /**
+      * Return true if bottom RelNode does not contain needToRetract and top RelNode need
+      * retraction from bottom RelNode. Currently, operators which contain aggregations need
+      * retraction from upstream nodes, besides, a needToRetract node also needs retraction from
+      * it's upstream nodes.
+      */
+    def bottomNeedToRetract(topRel: RelNode, bottomRel: RelNode): Boolean = {
+      val bottomTraits = bottomRel.getTraitSet
+      if(!traitSetContainNeedToRetract(bottomTraits)){
+        topRel match {
+          case _: DataStreamGroupAggregate => true
```

Comment:
I wonder whether it would be better to extend `DataStreamRel` with a couple of methods that expose the retraction behavior of each operator (for example `requiresUpdatesAsRetractions()`, `forwardsRetractions()`, `consumesRetractions()`, `producesRetractions()`, or `producesUpdates()`).

I think hardcoding the operator classes here is not a good solution, because it might easily break if the rules are not correctly updated when the operators change.